### DISCUSSION TITLE
[WinUI 3] COMException: The application called an interface that was marshalled for a different thread.

### DISCUSSION THREAD

**Discussion by Diyari-Kurdi** [microsoft/microsoft-ui-xaml#8410] [Q&A] [answered]
Hello,
I'm facing an issue using **SignalR** in **WinUI 3**. When SignalR fires an **event** and if I change a property inside that event while the property is bound to the View, it throws this exception:
_System.Runtime.InteropServices.COMException: 'The application called an interface that was marshalled for a different thread. (0x8001010E (RPC_E_WRONG_THREAD))'_

Source Code: 
[WinUI_SignalR_Demo.zip](https://github.com/microsoft/microsoft-ui-xaml/files/11285390/WinUI_SignalR_Demo.zip)

![image](https://user-images.githubusercontent.com/65598554/233348278-01df49d9-13af-4ac5-85c4-79aa4dd9a56a.png)

![image](https://user-images.githubusercontent.com/65598554/233348499-1cc2b544-b586-42e2-8606-035507d76d1a.png)

**Comment by kmgallahan** [comment, marked as answer]
The result of the GetMessage() is coming back on a non-UI thread, meaning PropertyChanged is invoked on a non-UI thread.

GPT-4 can handle the rest:

---

In WinUI or UWP, you cannot invoke PropertyChanged on a non-UI thread because the UI elements can only be accessed from the UI thread. If you attempt to access UI elements from a non-UI thread, you may encounter unexpected behavior or crashes in your application.

To update UI elements from a non-UI thread, you can use the `Dispatcher.RunAsync` method in UWP or `DispatcherQueue.TryEnqueue` method in WinUI 3. Here's an example of how to use `Dispatcher.RunAsync` in UWP:

```csharp
Windows.ApplicationModel.Core.CoreApplication.MainView.CoreWindow.Dispatcher.RunAsync(CoreDispatcherPriority.Normal,
() =>
{
    // update your UI here
    Update(i);
});
```
[Source 2](https://stackoverflow.com/questions/45129425/how-to-call-a-method-in-ui-thread-from-another-running-thread-in-uwp)

For WinUI 3, you can use the `DispatcherQueue.TryEnqueue` method:

```csharp
this.DispatcherQueue.TryEnqueue(() => 
{
    // update your UI here
    Update(i);
});
```
[Source 4](https://github.com/microsoft/microsoft-ui-xaml/issues/2609)

It's worth noting that in WinUI 3, the CoreDispatcher has been succeeded by DispatcherQueue. Most UI objects in WinUI 3 now have a DispatcherQueue property instead of a CoreDispatcher property. You can access the DispatcherQueue from any UI element or Window [Source 4](https://github.com/microsoft/microsoft-ui-xaml/issues/2609).

In summary, when updating UI elements from a non-UI thread, you should use `Dispatcher.RunAsync` in UWP or `DispatcherQueue.TryEnqueue` in WinUI 3 to ensure the changes are made on the UI thread to avoid any issues or crashes.
